portal/src/DetailsListWithOrdering.tsx

```diff
@@ -0,0 +1,125 @@
+import React from "react";
+import {
+  DetailsList,
+  IDetailsListProps,
+  IColumn,
+  IconButton,
+} from "@fluentui/react";
+import { Context } from "@oursky/react-messageformat";
+
+import styles from "./DetailsListWithOrdering.module.scss";
+
+interface DetailsListWithOrderingProps extends IDetailsListProps {
+  onSwapClicked: (swapUpward: boolean, index?: number) => void;
+  columns: IColumn[];
+  onRenderItemColumn: (
+    item?: any,
+    index?: number,
+    column?: IColumn
+  ) => React.ReactNode;
+  orderColumnMinWidth?: number;
+  orderColumnMaxWidth?: number;
+}
+
+interface OrderColumnButtonsProps {
+  index?: number;
+  onSwapClicked: (swapUpward: boolean, index?: number) => void;
+}
+
+const OrderColumnButtons: React.FC<OrderColumnButtonsProps> = function OrderColumnButtons(
+  props: OrderColumnButtonsProps
+) {
+  const onUpClicked = React.useCallback(() => {
+    props.onSwapClicked(true, props.index);
+  }, [props]);
+  const onDownClicked = React.useCallback(() => {
+    props.onSwapClicked(false, props.index);
+  }, [props]);
+  return (
+    <div>
+      <IconButton
+        className={styles.orderColumnButton}
+        onClick={onDownClicked}
+        iconProps={{ iconName: "ChevronDown" }}
+      />
+      <IconButton
+        className={styles.orderColumnButton}
+        onClick={onUpClicked}
+        iconProps={{ iconName: "ChevronUp" }}
+      />
+    </div>
+  );
+};
+
+export function useOnSwapClicked<T>(
+  state: T[],
+  setState: React.Dispatch<React.SetStateAction<T[]>>
+): (swapUpward: boolean, index?: number) => void {
+  const onSwapClicked = React.useCallback(
+    (swapUpward: boolean, index?: number) => {
+      if (index == null) {
+        return;
+      }
+      if (swapUpward && index > 0) {
+        setState((prev: T[]) => {
+          const target = prev[index - 1];
+          prev[index - 1] = prev[index];
+          prev[index] = target;
+          return [...prev];
+        });
+        return;
+      }
+      if (!swapUpward && index < state.length - 1) {
+        setState((prev: T[]) => {
+          const target = prev[index + 1];
+          prev[index + 1] = prev[index];
+          prev[index] = target;
+          return [...prev];
```

Comment:
Avoid mutation on the object.

Another example of making a copy first.

```typescript
const newState = [...prev];
// We can mutate newState because it is copied.
const thisObj = prev[index];
const thatObj = prev[index + 1];
newState[index + 1] = thisObj;
newState[index] = thatObj;
return newState;
```